src/ppx_deriving_qcheck/ppx_deriving_qcheck.ml

```diff
@@ -381,66 +351,123 @@ and gen_from_arrow ~loc ~env left right =
-        record ~loc ~gens xs
+        gen_record ~loc ~gens xs
     | _ ->
         let typ = Option.get td.ptype_manifest in
-        gen_from_type ~loc ~env ~typ_name:name typ
+        gen_from_type ~loc ~env typ
   in
-  let gen = curry_args ~loc args gen in
 
-  [%stri let [%p pat_gen] = [%e gen]]
+  let pat_gen = pat ~loc ty in
+  if not is_rec then
+    let gen = Args.curry_args ~loc args_pat gen in
+    `Normal [%stri let [%p pat_gen] = [%e gen]]
+  else
+    let gen = Args.curry_args ~loc (args_pat @ [A.pvar "n"]) gen in
+    let pat_gen_sized = pat ~loc ~sized:true ty in
+    let gen_sized = name ~sized:true ty |> A.evar in
+    let gen_normal =
+      Args.curry_args ~loc args_pat
+        (G.sized ~loc (Args.apply_args ~loc args_expr gen_sized))
+    in
+    `Recursive (
+        [%stri let rec [%p pat_gen_sized] = [%e gen]],
+        [%stri let [%p pat_gen] = [%e gen_normal]]
+      )
 
-let derive_gen ~loc xs : structure =
+(** *)
+let mutually_recursive_gens ~loc gens =
+  let (module A) = Ast_builder.make loc in
+  let to_mutualize_gens =
+    List.map (function
+        | `Recursive (x, _) -> x
+        | `Normal x -> x) gens
+  in
+  let normal_gens =
+    List.filter_map (function
+        | `Recursive (_, x) -> Some x
+        | `Normal _ -> None) gens
+  in
+  let gens =
+    List.map (function
+        | [%stri let [%p? pat] = [%e? expr]]
+          | [%stri let rec [%p? pat] = [%e? expr]] ->
+           A.value_binding ~pat ~expr
+        | _ -> assert false) to_mutualize_gens
+  in
+  let mutual_gens = A.pstr_value Recursive gens in
+  mutual_gens :: normal_gens
+
+(** [derive_gen ~loc xs] creates generators for type declaration in [xs].
+
+    It also has a hidden purpose: it sets the environment prior to
+    the derivation. It identifies recursive types declarations (e.g. list)
+    and its external arguments (e.g. 'a list).
+*)
+let derive_gen ~loc (xs : rec_flag * type_declaration list) : structure =
   match xs with
-  | _, [ x ] -> [ gen_from_type_declaration ~loc x ]
+  | (_, [ x ]) ->
+     let typ_name = x.ptype_name.txt in
+     let env = { curr_type = typ_name; rec_types = []; curr_types = [typ_name] } in
+     let env = add_if_rec env x typ_name in
+     (match gen_from_type_declaration ~loc ~env x with
+      | `Recursive (gen_sized, gen) -> [gen_sized; gen]
+      | `Normal gen -> [gen])
   | _, xs ->
-      let (module A) = Ast_builder.make loc in
-      let env =
-        List.fold_left
-          (fun env td ->
-            let x = td.ptype_name.txt in
-            let gen = name x |> A.evar in
-            let expr = [%expr [%e gen] ()] in
-            TypeGen.add x expr env)
-          TypeGen.empty xs
-      in
-      let gens = List.map (gen_from_type_declaration ~loc ~env) xs in
-      mutually_recursive_gens ~loc gens
+     let typ_names = List.map (fun x -> x.ptype_name.txt) xs in
+     let env = { curr_type = ""; rec_types = []; curr_types = typ_names } in
```

Comment:
`curr_type = ""` surprised me a bit at first.

In a sense it moves the "absence of `curr_type` representation from `None` to `""`.

However, we no longer have a `ref` - and here the absence is only local to `derive_gen`.
Any reader can thus convince him/her/themself that `curr_type` will be updated to be
non-empty just a few lines below (l.465) before being used.

I therefore think it is fine 👍‍ (and definitely better than the ref!)